### PR TITLE
net-vpn/i2pd: 2.25.0 add proxy-maint and bump EAPI

### PR DESCRIPTION
**Pull request by khumarahn**
Regular bump, no changes to 2.24.0. Old cleaned up to leave 3 latest versions.

@blueness please do not ignore.

### CODE REVIEW

**Comment by khumarahn** [comment]
Added sign-off and linked a bug.

**Comment by blueness** [comment]
pushed

**Comment by khumarahn** [comment]
Uh.... I added empty lines in the commit messages. Was this the problem?

I cannot say anything intelligent about gentoo-proxy-maintainers.

**Comment by blueness** [comment]
yeah something weird happened with the commit messages.

anyhow, i don't really want responsibility for this package anymore.  i've dropped myself from the metadata.xml.

**Comment by mgorny** [comment]
1. EAPI 7, please.
2. If you want proxy-maint to commit for you, you need to add us to metadata.xml.

**Comment by khumarahn** [comment]
I am puzzled by eapi 7. I set `EAPI=7` and remove `inherit eapi7-ver`. From what I understand, no other changes should be necessary here. But then the build fails with
```
/usr/bin/x86_64-pc-linux-gnu-g++  -O2 -pipe  -march=native -Wall -Wextra -Winvalid-pch -Wno-unused-parameter -std=c++11 -pipe -maes -mavx -pthread -fPIC  -Wl,-O1 -Wl,--as-needed -rdynamic CMakeFiles/i2pd.dir/var/tmp/portage/net-vpn/i2pd-2.25.0/work/i2pd-2.25.0/daemon/Daemon.cpp.o CMakeFiles/i2pd.dir/var/tmp/portage/net-vpn/i2pd-2.25.0/work/i2pd-2.25.0/daemon/HTTPServer.cpp.o CMakeFiles/i2pd.dir/var/tmp/portage/net-vpn/i2pd-2.25.0/work/i2pd-2.25.0/daemon/I2PControl.cpp.o CMakeFiles/i2pd.dir/var/tmp/portage/net-vpn/i2pd-2.25.0/work/i2pd-2.25.0/daemon/i2pd.cpp.o CMakeFiles/i2pd.dir/var/tmp/portage/net-vpn/i2pd-2.25.0/work/i2pd-2.25.0/daemon/UPnP.cpp.o CMakeFiles/i2pd.dir/var/tmp/portage/net-vpn/i2pd-2.25.0/work/i2pd-2.25.0/daemon/UnixDaemon.cpp.o  -o i2pd -Wl,-rpath,/var/tmp/portage/net-vpn/i2pd-2.25.0/work/i2pd-2.25.0_build: -lminiupnpc libi2pdclient.so -ldl -lboost_system-mt -lboost_filesystem-mt -lboost_program_options-mt -lboost_date_time-mt -lssl -lcrypto -lz -pthread -ldl -lboost_system-mt -lboost_filesystem-mt -lboost_program_options-mt -lboost_date_time-mt -lssl -lcrypto -lz libi2pd.so                                                                                  
/usr/lib/gcc/x86_64-pc-linux-gnu/8.3.0/../../../../x86_64-pc-linux-gnu/bin/ld: libi2pd.so: undefined reference to `boost::gregorian::greg_month::as_short_string() const'
/usr/lib/gcc/x86_64-pc-linux-gnu/8.3.0/../../../../x86_64-pc-linux-gnu/bin/ld: libi2pd.so: undefined reference to `boost::gregorian::greg_month::get_month_map_ptr[abi:cxx11]()'
collect2: error: ld returned 1 exit status
```

One immediate difference I notice is that in EAPI 7, ninja is used by default instead of cmake. But even with `CMAKE_MAKEFILE_GENERATOR="emake"`, the build fails with the same problem.

With `EAPI=6`, the build completes. Any pointers?

**Comment by khumarahn** [comment]
The 2.25.0 version bump is now in the tree, so I am now re-basing this pull request off the current tree, aiming at eapi-7.

It would be good to fix the previous commit names, but I do not now how to do it.

**Comment by khumarahn** [comment]
Another difference: with EAPI=7, cmake is building a shared library `libi2pdclient.so`. Instead, with EAPI=6 this is a static `libi2pdclient.a`.

**Comment by khumarahn** [comment]
I think I found the problem. In `EAPI=7`, `BUILD_SHARED_LIBS=ON` by default. With this, i2pd fails. Reverting back to OFF solves build problems. I added commit with EAPI bump. Please review.

**Comment by mgorny** [comment]
Report this upstream. They shouldn't be randomly failing on build configuration.

**Comment by khumarahn** [comment]
If I want to look at what options cmake uses in gentoo ebuild, what is the best way for this?

**Comment by khumarahn** [comment]
If I build 2.25.0 in a terminal with
```
cmake -DBUILD_SHARED_LIBS=ON -DWITH_AESNI=ON -DWITH_AVX=ON -DWITH_UPNP=ON
make
```
then the build succeeds. Apparently there is some gentoo config other than `BUILD_SHARED_LIBS=ON` which is activated in EAPI=7 and makes the build fail.

**Comment by mgorny** [comment]
May be related to linker in use. The underlying problem seems to be that the library is not linked to boost, and upstream instead relies on linking the final program to it.

**Comment by khumarahn** [comment]
I know more now. The minimal way to reproduce the problem is
```
export LDFLAGS="-Wl,--as-needed"
cmake -DBUILD_SHARED_LIBS=ON .
make
```
It does not happen withour either `BUILD_SHARED_LIBS=ON` or `-Wl,--as-needed`.

I will talk to upstream now.

**Comment by khumarahn** [comment]
The issue with `--as-needed` has been fixed upstream by linking libi2pd to boost and zlib, see
https://github.com/PurpleI2P/i2pd/issues/1353
We now use that patch for 2.25.0, and it won't be needed in future releases.

There is another problem that install path for 64bit libraries is /usr/lib instead of /usr/lib64. I added a patch that takes care of this. Not sure what would be the long term fix here.

Please review.

File: net-vpn/i2pd/files/i2pd-2.25.0-lib-path.patch [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+index 949f6a46..42165524 100644
+--- a/build/CMakeLists.txt
++++ b/build/CMakeLists.txt
+@@ -103,8 +103,8 @@ set_target_properties(libi2pd PROPERTIES PREFIX "")
+ if (WITH_LIBRARY)
+   install(TARGETS libi2pd
+     EXPORT libi2pd
+-    ARCHIVE DESTINATION lib
+-    LIBRARY DESTINATION lib
++    #ARCHIVE DESTINATION lib
```
Comment (mgorny) [inline net-vpn/i2pd/files/i2pd-2.25.0-lib-path.patch]: Don't comment things out in a patch-hack. Just remove them altogether.

Also, have you reported the problem upstream?

File: net-vpn/i2pd/files/i2pd-2.25.0-link.patch [hunk 2]
```diff
@@ -0,0 +1,12 @@
+diff --git a/build/CMakeLists.txt b/build/CMakeLists.txt
```
Comment (mgorny) [inline net-vpn/i2pd/files/i2pd-2.25.0-link.patch]: Likewise, please link upstream bug report.

File: net-vpn/i2pd/i2pd-2.25.0.ebuild [hunk 3]
```diff
@@ -43,7 +43,9 @@ CMAKE_USE_DIR="${S}/build"
 
 DOCS=( README.md contrib/i2pd.conf contrib/tunnels.conf )
 
-PATCHES=( "${FILESDIR}/${PN}-2.14.0-fix_installed_components.patch" )
+PATCHES=( "${FILESDIR}/${PN}-2.14.0-fix_installed_components.patch"
+	"${FILESDIR}/i2pd-2.25.0-link.patch"
```
Comment (mgorny) [inline net-vpn/i2pd/i2pd-2.25.0.ebuild]: Does an EAPI bump actually imply all those patches? Or have you added independent fixes?

**Comment by mgorny** [comment]
That said, there's a trick you may want to use: if you move fowners/fperms from `src_install` to `pkg_preinst`, you can also delay user/group creation until that phase.

File: net-vpn/i2pd/metadata.xml [hunk 4]
```diff
@@ -1,6 +1,10 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
 <pkgmetadata>
+	<maintainer type="project">
```
[the remark this answers is not in the document]
Comment (khumarahn) [reply net-vpn/i2pd/metadata.xml]: ok, I changed this. Is the new order OK?

File: net-vpn/i2pd/files/i2pd-2.25.0-lib-path.patch (hunk 1 above)
Comment (khumarahn) [reply net-vpn/i2pd/files/i2pd-2.25.0-lib-path.patch]: Deleted. Did not report, I don't yet understand myself what should be there instead.

File: net-vpn/i2pd/files/i2pd-2.25.0-link.patch (hunk 2 above)
Comment (khumarahn) [reply net-vpn/i2pd/files/i2pd-2.25.0-link.patch]: linked

**Comment by khumarahn** [comment]
Thanks for the advice, I will look at  fowners/fperms when this PR is done.

**Comment by mgorny** [comment]
Revbump for potentially breaking changes.

**Comment by khumarahn** [comment]
I don't expect more problems, but the revbump is reasonable. Changed.

**Pull request closed without merging**